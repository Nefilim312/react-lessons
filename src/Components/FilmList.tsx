import React from 'react';
import { IFilm } from '../interfaces';
import FilmTile from './FilmTile';
import { createUseStyles } from 'react-jss';

const useStyles = createUseStyles({
  filmList: {
    background: 'var(--backGroundGrey)',
    display: 'grid',
    gridTemplateColumns: '1fr 1fr 1fr',
    gridGap: '3rem',
    padding: '4rem 3rem',
  },
  empty: {
    fontSize: '4rem',
    padding: '10rem',
    fontWeight: '300',
    color: '#AFAFAF',
    textAlign: 'center',
    background: 'var(--backGroundGrey)',
  },
});

interface FilmListProps {
  filmSet: IFilm[];
  onItemClick: (item: IFilm) => void;
}

const FilmList: React.FC<FilmListProps> = ({
  filmSet,
  onItemClick,
}: FilmListProps) => {
  const classes = useStyles();

  if (filmSet.length) {
    return (
      <div className={classes.filmList}>
        {filmSet.map((film, idx) => (
          <FilmTile onClick={onItemClick} filmData={film} key={idx}></FilmTile>
        ))}
      </div>
    );
  } else {
    return <div className={classes.empty}>No films found</div>;
  }
};

export default FilmList;
