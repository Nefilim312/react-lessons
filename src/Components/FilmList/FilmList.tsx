import React from 'react';
import FilmTile from '../FilmTile/FilmTile';
import Loader from '../Loader/Loader';
import { createUseStyles } from 'react-jss';
import { BACKGROUND } from '../../constants/styles';

interface FilmListProps {
  films?: IFilm[];
  loading?: boolean;
}

const FilmList: React.FC<FilmListProps> = ({ films, loading }) => {
  const classes = useStyles();

  if (loading) {
    return <Loader />;
  }

  if (films?.length) {
    return (
      <div className={classes.filmList}>
        {films.map((film) => (
          <FilmTile key={film.id} filmData={film} />
        ))}
      </div>
    );
  } else {
    return <div className={classes.empty}>No films found</div>;
  }
};

const useStyles = createUseStyles({
  filmList: {
    background: BACKGROUND,
    display: 'grid',
    gridTemplateColumns: '1fr 1fr 1fr',
    gridGap: '3rem',
    padding: '4rem 3rem',
  },
  empty: {
    fontSize: '4rem',
    padding: '10rem',
    fontWeight: '300',
    color: '#AFAFAF',
    textAlign: 'center',
    background: BACKGROUND,
  },
});

export default FilmList;
