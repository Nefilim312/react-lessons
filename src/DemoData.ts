import { IFilm } from './interfaces';

const DemoData: IFilm[] = [
  {
    title: 'Four Rooms',
    img:
      'https://m.media-amazon.com/images/M/MV5BNDc3Y2YwMjUtYzlkMi00MTljLTg1ZGMtYzUwODljZTI1OTZjXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_.jpg',
    date: '1995',
    description:
      "Four interlocking tales that take place in a fading hotel on New Year's Eve.",
    genre: 'Comedy',
    length: '98',
    rating: '6.8',
  },
  {
    title: 'Jackie Brown',
    img:
      'https://m.media-amazon.com/images/M/MV5BNmY5ODRmYTItNWU0Ni00MWE3LTgyYzUtYjZlN2Q5YTcyM2NmXkEyXkFqcGdeQXVyNjU0OTQ0OTY@._V1_SY1000_CR0,0,683,1000_AL_.jpg',
    date: '1997',
    description:
      'A middle-aged woman finds herself in the middle of a huge conflict that will either make her a profit or cost her life.',
    genre: 'Crime, Drama, Thriller',
    length: '154',
    rating: '7.5',
  },
  {
    title: 'Kill Bill: Vol. 1',
    img:
      'https://m.media-amazon.com/images/M/MV5BNzM3NDFhYTAtYmU5Mi00NGRmLTljYjgtMDkyODQ4MjNkMGY2XkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_.jpg',
    date: '2003',
    description:
      'After awakening from a four-year coma, a former assassin wreaks vengeance on the team of assassins who betrayed her.',
    genre: 'Action, Crime, Thriller',
    length: '111',
    rating: '8.1',
  },
  {
    title: 'Pulp Fiction',
    img:
      'https://m.media-amazon.com/images/M/MV5BNGNhMDIzZTUtNTBlZi00MTRlLWFjM2ItYzViMjE3YzI5MjljXkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_SY1000_CR0,0,686,1000_AL_.jpg',
    date: '1994',
    description:
      'The lives of two mob hitmen, a boxer, a gangster and his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.',
    genre: 'Crime, Drama',
    length: '154',
    rating: '8.9',
  },
  {
    title: 'Reservoir Dogs',
    img:
      'https://m.media-amazon.com/images/M/MV5BZmExNmEwYWItYmQzOS00YjA5LTk2MjktZjEyZDE1Y2QxNjA1XkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_.jpg',
    date: '1992',
    description:
      'When a simple jewelry heist goes horribly wrong, the surviving criminals begin to suspect that one of them is a police informant.',
    genre: 'Crime, Drama, Thriller',
    length: '94',
    rating: '8.3',
  },
  {
    title: 'Kill Bill: Vol. 2',
    img:
      'https://m.media-amazon.com/images/M/MV5BNmFiYmJmN2QtNWQwMi00MzliLThiOWMtZjQxNGRhZTQ1MjgyXkEyXkFqcGdeQXVyNzQ1ODk3MTQ@._V1_.jpg',
    date: '2004',
    description:
      'The Bride continues her quest of vengeance against her former boss and lover Bill, the reclusive bouncer Budd, and the treacherous, one-eyed Elle.',
    genre: 'Action, Crime, Thriller',
    length: '137',
    rating: '8.0',
  },
];

export default DemoData;
