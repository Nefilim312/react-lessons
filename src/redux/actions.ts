import {
  SHOW_LOADER,
  HIDE_LOADER,
  REQUEST_MOVIES,
  CHANGE_FILTER,
  LOAD_FILM,
  REQUEST_SIMILAR_MOVIES,
} from './types';
import { Dispatch } from 'redux';
import fetch from 'node-fetch';

export function requestMovies(filter?: IFilter, needShowLoader = true) {
  return async (dispatch: Dispatch) => {
    if (!filter?.search) {
      return dispatch({ type: REQUEST_MOVIES, payload: [] });
    }

    needShowLoader && dispatch(showLoader());

    const url = new URL('https://reactjs-cdp.herokuapp.com/movies');
    const searchObj: { [key: string]: any } = {
      sortOrder: 'desc',
      limit: '9',
      ...filter,
    };

    const searchParams = new URLSearchParams(searchObj);

    url.search = searchParams.toString();

    const response = await fetch(url.toString());
    const json = await response.json();

    needShowLoader && dispatch(hideLoader());
    return dispatch({ type: REQUEST_MOVIES, payload: json.data });
  };
}

export function changeFilter(filter: IFilter) {
  return { type: CHANGE_FILTER, payload: filter };
}

export function resetFilter() {
  return {
    type: CHANGE_FILTER,
    payload: { search: '', searchBy: 'title', sortBy: 'release_date' },
  };
}

export function showLoader() {
  return { type: SHOW_LOADER };
}

export function hideLoader() {
  return { type: HIDE_LOADER };
}

export function loadFilm(filmId: number) {
  return async (dispatch: Dispatch) => {
    const url = new URL(`https://reactjs-cdp.herokuapp.com/movies/${filmId}`);

    const response = await fetch(url.toString());
    const data = await response.json();

    return dispatch({ type: LOAD_FILM, payload: data });
  };
}

export function requestSimilarMovies(genres: string[]) {
  return async (dispatch: Dispatch) => {
    dispatch(showLoader());

    const url = new URL('https://reactjs-cdp.herokuapp.com/movies');
    const searchObj: { [key: string]: any } = {
      sortOrder: 'desc',
      limit: '3',
      sortBy: 'release_date',
      filter: genres,
    };

    const searchParams = new URLSearchParams(searchObj);

    url.search = searchParams.toString();

    const response = await fetch(url.toString());
    const json = await response.json();

    dispatch(hideLoader());
    return dispatch({ type: REQUEST_SIMILAR_MOVIES, payload: json.data });
  };
}
