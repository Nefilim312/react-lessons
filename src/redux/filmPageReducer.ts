import {
  REQUEST_SIMILAR_MOVIES,
  LOAD_FILM,
  SHOW_LOADER,
  HIDE_LOADER,
  ActionTypes,
} from './types';
import { HYDRATE } from 'next-redux-wrapper';

const initialState: IFilmPageState = {
  movies: [],
  film: {
    id: 0,
    title: '',
    poster_path: '',
    release_date: '',
    overview: '',
    genres: [],
    runtime: '',
    vote_average: '',
  },
  loading: false,
};

export function filmPageReducer(
  state = initialState,
  action: ActionTypes
): IFilmPageState {
  switch (action.type) {
    case REQUEST_SIMILAR_MOVIES:
      return { ...state, movies: action.payload };
    case LOAD_FILM:
      return { ...state, film: action.payload };
    case SHOW_LOADER:
      return { ...state, loading: true };
    case HIDE_LOADER:
      return { ...state, loading: false };
    case HYDRATE:
      return {
        ...state,
        ...action.payload.filmPage,
      };
    default:
      return state;
  }
}

export type FilmPageState = ReturnType<typeof filmPageReducer>;

export default filmPageReducer;
