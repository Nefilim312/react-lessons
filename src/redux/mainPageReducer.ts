import {
  REQUEST_MOVIES,
  CHANGE_FILTER,
  SHOW_LOADER,
  HIDE_LOADER,
  ActionTypes,
} from './types';
import { HYDRATE } from 'next-redux-wrapper';

export const initialState: IMainPageState = {
  movies: [],
  filter: { search: '', searchBy: 'title', sortBy: 'release_date' },
  loading: false,
};

export function mainPageReducer(
  state = initialState,
  action: ActionTypes
): IMainPageState {
  switch (action.type) {
    case REQUEST_MOVIES:
      return { ...state, movies: action.payload };
    case CHANGE_FILTER:
      return { ...state, filter: action.payload };
    case SHOW_LOADER:
      return { ...state, loading: true };
    case HIDE_LOADER:
      return { ...state, loading: false };
    case HYDRATE:
      return {
        ...state,
        ...action.payload.mainPage,
      };
    default:
      return state;
  }
}

export type RootState = ReturnType<typeof mainPageReducer>;

export default mainPageReducer;
