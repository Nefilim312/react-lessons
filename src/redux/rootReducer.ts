import {
  REQUEST_MOVIES,
  CHANGE_FILTER,
  SHOW_LOADER,
  HIDE_LOADER,
  ActionTypes,
} from './types';
import { REHYDRATE } from 'redux-persist';

const initialState: IState = {
  movies: [],
  filter: { search: '', searchBy: 'title', sortBy: 'release_date' },
  loading: false,
};

export function rootReducer(state = initialState, action: ActionTypes): IState {
  switch (action.type) {
    case REQUEST_MOVIES:
      return { ...state, movies: action.payload };
    case CHANGE_FILTER:
      return { ...state, filter: action.payload };
    case SHOW_LOADER:
      return { ...state, loading: true };
    case HIDE_LOADER:
      return { ...state, loading: false };
    case REHYDRATE:
      return {
        ...state,
        movies: action.payload.movies,
        filter: action.payload.filter,
      };
    default:
      return state;
  }
}

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
