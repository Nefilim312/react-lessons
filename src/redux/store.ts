import { createStore, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from './rootReducer';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['filter'],
};
const persistedReducer = persistReducer(persistConfig, rootReducer);
const store = createStore(persistedReducer, compose(applyMiddleware(thunk)));
const persistor = persistStore(store);

export { store, persistor };
