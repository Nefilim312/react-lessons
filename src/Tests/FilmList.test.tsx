import React from 'react';
import { mount } from 'enzyme';
import FilmList from '../Components/FilmList';
import DemoData from './DemoData';
import { Provider } from 'react-redux';
import { store } from '../redux/store';

describe('FilmList component', () => {
  it('should be render text "No films found"', () => {
    const component = mount(
      <Provider store={store}>
        <FilmList films={[]} onItemClick={() => {}} />
      </Provider>
    );

    expect(component.find('div')).toMatchSnapshot();
  });

  it('should be render demo data (6 elements)', () => {
    const component = mount(
      <Provider store={store}>
        <FilmList films={DemoData} onItemClick={() => {}} />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  it('test click event', () => {
    const mockCallback = jest.fn();
    const filmList = mount(
      <Provider store={store}>
        <FilmList films={DemoData} onItemClick={mockCallback} />
      </Provider>
    );

    filmList.find(FilmList).childAt(0).childAt(0).simulate('click');
    expect(mockCallback).toBeCalled();
  });
});
