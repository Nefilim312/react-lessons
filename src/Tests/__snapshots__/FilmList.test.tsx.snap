// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmList component should be render demo data (6 elements) 1`] = `
<div
  className="filmList-0-2-1"
>
  <FilmTile
    filmData={
      Object {
        "date": "1995",
        "description": "Four interlocking tales that take place in a fading hotel on New Year's Eve.",
        "genre": "Comedy",
        "img": "https://m.media-amazon.com/images/M/MV5BNDc3Y2YwMjUtYzlkMi00MTljLTg1ZGMtYzUwODljZTI1OTZjXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_.jpg",
        "length": "98",
        "rating": "6.8",
        "title": "Four Rooms",
      }
    }
    key="0"
    onClick={[Function]}
  />
  <FilmTile
    filmData={
      Object {
        "date": "1997",
        "description": "A middle-aged woman finds herself in the middle of a huge conflict that will either make her a profit or cost her life.",
        "genre": "Crime, Drama, Thriller",
        "img": "https://m.media-amazon.com/images/M/MV5BNmY5ODRmYTItNWU0Ni00MWE3LTgyYzUtYjZlN2Q5YTcyM2NmXkEyXkFqcGdeQXVyNjU0OTQ0OTY@._V1_SY1000_CR0,0,683,1000_AL_.jpg",
        "length": "154",
        "rating": "7.5",
        "title": "Jackie Brown",
      }
    }
    key="1"
    onClick={[Function]}
  />
  <FilmTile
    filmData={
      Object {
        "date": "2003",
        "description": "After awakening from a four-year coma, a former assassin wreaks vengeance on the team of assassins who betrayed her.",
        "genre": "Action, Crime, Thriller",
        "img": "https://m.media-amazon.com/images/M/MV5BNzM3NDFhYTAtYmU5Mi00NGRmLTljYjgtMDkyODQ4MjNkMGY2XkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_.jpg",
        "length": "111",
        "rating": "8.1",
        "title": "Kill Bill: Vol. 1",
      }
    }
    key="2"
    onClick={[Function]}
  />
  <FilmTile
    filmData={
      Object {
        "date": "1994",
        "description": "The lives of two mob hitmen, a boxer, a gangster and his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.",
        "genre": "Crime, Drama",
        "img": "https://m.media-amazon.com/images/M/MV5BNGNhMDIzZTUtNTBlZi00MTRlLWFjM2ItYzViMjE3YzI5MjljXkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_SY1000_CR0,0,686,1000_AL_.jpg",
        "length": "154",
        "rating": "8.9",
        "title": "Pulp Fiction",
      }
    }
    key="3"
    onClick={[Function]}
  />
  <FilmTile
    filmData={
      Object {
        "date": "1992",
        "description": "When a simple jewelry heist goes horribly wrong, the surviving criminals begin to suspect that one of them is a police informant.",
        "genre": "Crime, Drama, Thriller",
        "img": "https://m.media-amazon.com/images/M/MV5BZmExNmEwYWItYmQzOS00YjA5LTk2MjktZjEyZDE1Y2QxNjA1XkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_.jpg",
        "length": "94",
        "rating": "8.3",
        "title": "Reservoir Dogs",
      }
    }
    key="4"
    onClick={[Function]}
  />
  <FilmTile
    filmData={
      Object {
        "date": "2004",
        "description": "The Bride continues her quest of vengeance against her former boss and lover Bill, the reclusive bouncer Budd, and the treacherous, one-eyed Elle.",
        "genre": "Action, Crime, Thriller",
        "img": "https://m.media-amazon.com/images/M/MV5BNmFiYmJmN2QtNWQwMi00MzliLThiOWMtZjQxNGRhZTQ1MjgyXkEyXkFqcGdeQXVyNzQ1ODk3MTQ@._V1_.jpg",
        "length": "137",
        "rating": "8.0",
        "title": "Kill Bill: Vol. 2",
      }
    }
    key="5"
    onClick={[Function]}
  />
</div>
`;

exports[`FilmList component should be render text "No films found" 1`] = `
<div
  className="empty-0-2-2"
>
  No films found
</div>
`;
